appscale/tools/node_layout.py

```diff
@@ -590,10 +597,46 @@ def is_database_replication_valid(self, nodes):
         self.replication = database_node_count
 
     if self.replication > database_node_count:
-      return self.invalid("Replication factor cannot exceed # of databases")
+      self.invalid("Replication factor cannot exceed # of databases.")
 
-    return self.valid()
+    return True
+
+  def distribute_unassigned_roles(self, nodes, role_count, master_node):
+    """ Distributes roles that were not defined by user.
+    
+    Args:
+      nodes: The list of nodes.
+      role_count: A dict containing roles mapped to their count.
+      master_node: The master node of the deployment.
+    """
+    # Check if a master node was specified.
+    if role_count.pop('master', 0) == 0:
```

Comment:
If that's the case, you can ignore it in the loop. Altering a data structure that's been given can be surprising to the calling function.